src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/data/SkillProvider.java

```diff
@@ -18,58 +17,50 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Objects;
 import java.util.Set;
 import java.util.function.Consumer;
 
 /**
- * Data provider for skill jsons
+ * Base class for skill data generators.
  */
 public abstract class SkillProvider implements DataProvider {
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
     private static final Logger LOGGER = LogManager.getLogger();
-    private static final Gson GSON = (new GsonBuilder()).setPrettyPrinting().create();
+    private final DataGenerator generator;
+    private final String namespace;
+    private Set<ResourceLocation> data;
 
-    private final DataGenerator      generator;
-    private final String             namespace;
-    private Set<ResourceLocation> ids;
-
-    /**
-     * Create a skill provider for the given namespace
-     *
-     * @param generator the data generator
-     * @param namespace the namespace to use in data generation
-     */
-    protected SkillProvider(DataGenerator generator, String namespace) {
-        this.generator = generator;
+    protected SkillProvider(String namespace, DataGenerator generator) {
         this.namespace = namespace;
+        this.generator = generator;
     }
 
+    protected abstract void createSkills(Consumer<SkillBuilder> consumer);
+
     @Internal
     @Override
-    public final void run(HashCache pCache) {
-        Path path = generator.getOutputFolder();
-        ids = Sets.newHashSet();
+    public void run(HashCache pCache) {
+        data = new HashSet<>();
         createSkills(skill -> {
-            if (!ids.add(skill.getId())) throw new IllegalStateException("Duplicate skill " + skill.getId());
+            if (!data.add(skill.getId())) throw new IllegalStateException("Duplicate skill " + skill.getId());
             else {
-                saveSkill(pCache, skill.serialize(), path.resolve("data/" + skill.getId().getNamespace() + "/am_skills/" + skill.getId().getPath() + ".json"));
+                saveSkill(pCache, skill.serialize(), generator.getOutputFolder().resolve("data/" + skill.getId().getNamespace() + "/am_skills/" + skill.getId().getPath() + ".json"));
             }
         });
     }
 
-    public Set<ResourceLocation> getSkills() {
-        return Collections.unmodifiableSet(ids);
+    @Override
+    public String getName() {
+        return "Skills";
```

Comment:
The name should be different for each instance with a different output namespace